core/upgrade/src/main/java/org/opennms/upgrade/implementations/VmwareAssetsMigratorOffline.java

```diff
@@ -0,0 +1,176 @@
+            final Statement preExecutionStatement = connection.createStatement();
+            try (final ResultSet preExecutionResultSet = preExecutionStatement.executeQuery("SELECT EXISTS (SELECT 1 FROM pg_attribute WHERE attrelid = (SELECT oid FROM pg_class WHERE relname = 'assets') AND attname = 'vmwaretopologyinfo')")) {
+                preExecutionResultSet.next();
+                if (!preExecutionResultSet.getBoolean(1)) {
+                    throw new OnmsUpgradeException("The 'vmwaretopologyinfo' columns do not exist anymore");
+                }
+            } catch (SQLException e) {
+                e.printStackTrace();
+                throw new OnmsUpgradeException("Error checking for column 'vmwaretopologyinfo'", e);
+            }
+        } catch (SQLException e) {
+            throw new OnmsUpgradeException("Error opening database connection", e);
+        }
+    }
+
+    @Override
+    public void postExecute() throws OnmsUpgradeException {
+    }
+
+    @Override
+    public void rollback() throws OnmsUpgradeException {
+    }
+
+    @Override
+    public void execute() throws OnmsUpgradeException {
+        long nodeCount = 0;
+
+        try (final Connection connection = DataSourceFactory.getInstance().getConnection()) {
+            connection.setAutoCommit(false);
+
+            do {
+                try (final Statement selectStatement = connection.createStatement(); final ResultSet resultSet = selectStatement.executeQuery("SELECT nodeid, vmwaremanagedobjectid, vmwaremangedentitytype, vmwaremanagementserver, vmwaretopologyinfo, vmwarestate FROM assets WHERE vmwaremanagementserver IS NOT NULL LIMIT " + BATCH_SIZE)) {
+
+                    if (!resultSet.next()) {
+                        break;
+                    }
+
+                    try (final PreparedStatement insertStatement = connection.prepareStatement("INSERT INTO node_metadata (id, context, key, value) VALUES  (?,?,?,?)");
+                         final PreparedStatement nullifyStatement = connection.prepareStatement("UPDATE assets SET vmwaremanagedobjectid=NULL, vmwaremangedentitytype=NULL, vmwaremanagementserver=NULL, vmwaretopologyinfo=NULL, vmwarestate=NULL WHERE nodeid=?")) {
+
+                        do {
+                            final Integer nodeId = resultSet.getInt("nodeid");
+                            final String vmwareManagedObjectId = resultSet.getString("vmwaremanagedobjectid");
+                            final String vmwareMangedentityType = resultSet.getString("vmwaremangedentitytype");
+                            final String vmwareManagementServer = resultSet.getString("vmwaremanagementserver");
+                            final String vmwareTopologyInfo = resultSet.getString("vmwaretopologyinfo");
+                            final String vmwareState = resultSet.getString("vmwarestate");
+
+                            insertStatement.setInt(1, nodeId);
+                            insertStatement.setString(2, "VMware");
```

Comment:
Can you reference the constants from `VmwareTopologyProvider` without entering dependency hell?

Reply:
My idea was to reference all constants from the VmwareImporter. I tried, but failed so hard. Topology is really special concerning dependencies.